File: pipeline/web/static/public/js/pages/doctables.js
/* Document table extraction viewer (BETA-099).
 *
 * Tables detected in a parsed document — the grid the parser produced, its
 * page context, its extraction status, and a CSV of exactly those cells. No
 * cell is re-detected; the source document is the authority for anything the
 * parse got wrong.
 *
 * State is the query: `#/doctables?doc=<document_id>` or `&table=<table_id>`.
 */
'use strict';

import { el, replace, fetchJSON, isoDate, sourceLink, num } from '/app.js';
import { section, pinnedCaveat, noData, errorCard } from '/js/components.js';

const STATUS_BADGE = { structured: 'ok', markdown_only: 'unverified', empty: 'muted' };

function readQuery(params) {
  const q = params || new URLSearchParams(location.hash.split('?')[1] || '');
  return { doc: q.get('doc') || '', table: q.get('table') || '' };
}

function toCSV(grid) {
  return grid.map((row) => row.map((c) => {
    const s = String(c ?? '');
    return /[",\n]/.test(s) ? `"${s.replace(/"/g, '""')}"` : s;
  }).join(',')).join('\r\n');
}

function download(name, text, type) {
  try {
    const url = URL.createObjectURL(new Blob([text], { type }));
    const a = el('a', { href: url, download: name });
    document.body.appendChild(a); a.click(); a.remove();
    setTimeout(() => URL.revokeObjectURL(url), 1000);
  } catch (e) { /* the browser blocked it */ }
}

function gridTable(grid) {
  if (!grid.length) return el('p', { class: 'small muted', text: 'No structured grid was extracted for this table.' });
  const [head, ...body] = grid;
  return el('div', { class: 'dt-scroll' }, el('table', { class: 'dt-grid' },
    el('thead', {}, el('tr', {}, ...head.map((c) => el('th', { text: c })))),
    el('tbody', {}, ...body.map((r) => el('tr', {}, ...r.map((c) => el('td', { text: c })))))));
}

async function renderDetail(main, params, tableId, backHref) {
  let data;
  try { data = await fetchJSON('document_tables', { table_id: tableId }); }
  catch (error) {
    replace(main, el('div', { class: 'section' }, errorCard(error, () => renderDetail(main, params, tableId, backHref))));
    return;
  }
  const page = el('div', {},
    el('div', { class: 'hero' },
      el('p', {}, el('a', { href: backHref, text: '← All tables in this document' })),
      el('h1', { text: `Table on page ${data.page_number ?? '—'}` }),
      data.caption ? el('p', { class: 'lede', text: data.caption }) : null,
      el('p', { class: 'small muted' },
        el('span', { class: `badge ${STATUS_BADGE[data.extraction_status] || 'muted'}`, text: data.extraction_status }),
        el('span', { text: ` ${num(data.row_count)} × ${num(data.column_count)} · ` }),
        el('a', { href: data.reading_room_link, text: 'open in the reading room' }))),
    el('div', { class: 'panel' },
      el('div', { class: 'dt-actions' },
        el('button', { class: 'btn', type: 'button',
          onclick: () => download(`${data.document_table_id}.csv`, toCSV(data.grid), 'text/csv') },
          'Download CSV'),
        data.document.source_url ? sourceLink(data.document.source_url, 'the source document ↗') : null),
      gridTable(data.grid),
      data.markdown ? el('details', { class: 'dt-md' },
        el('summary', { text: 'Parser markdown' }),
        el('pre', { class: 'small', text: data.markdown })) : null,
      pinnedCaveat(data.note, 'Read this with the grid')),
    data.context.length ? el('div', { class: 'panel' },
      section('Around this table', null,
        el('ul', { class: 'small' }, ...data.context.map((c) => el('li', {},
          el('span', { class: 'muted', text: `${c.element_type || 'text'}: ` }), c.text))))) : null);
  replace(main, page);
}

export async function render(main, { params = null } = {}) {
  const cur = readQuery(params);

  const docInput = el('input', { value: cur.doc, class: 'dt-input',
    placeholder: 'document_id', 'aria-label': 'document id' });
  const controls = el('div', { class: 'panel dt-controls' }, docInput,
    el('button', { class: 'btn primary', type: 'button',
      onclick: () => { location.hash = docInput.value.trim() ? `#/doctables?doc=${encodeURIComponent(docInput.value.trim())}` : '#/doctables'; } },
      'Show tables'));

  if (cur.table && cur.doc) {
    await renderDetail(main, params, cur.table, `#/doctables?doc=${encodeURIComponent(cur.doc)}`);
    return () => {};
  }

  const page = el('div', {},
    el('div', { class: 'hero' },
      el('h1', { text: 'Document tables' }),
      el('p', { class: 'lede', text:
        'Tables detected in a parsed document, shown as the grid the parser '
        + 'produced. No cell is re-detected — the source document is the '
        + 'authority for anything the parse got wrong.' })),
    controls);

  if (!cur.doc) {
    page.append(el('div', { class: 'panel' },
      pinnedCaveat('Give a document_id (from document search).', 'How to use this')));
    replace(main, page);
    return () => {};
  }

  let data;
  try { data = await fetchJSON('document_tables', { document_id: cur.doc }); }
  catch (error) {
    page.append(el('div', { class: 'section' }, errorCard(error, () => render(main, { params }))));
    replace(main, page);
    return () => {};
  }

  const byStatus = data.counts.by_status || {};
  page.append(el('div', { class: 'panel' },
    el('h2', { text: data.document.title || data.document.document_id }),
    el('p', { class: 'small muted' },
      data.document.source_url ? sourceLink(data.document.source_url, 'source ↗') : null,
      el('span', { text: ` retrieved ${isoDate(data.document.retrieved_at)} · `
        + data.statuses.filter((s) => byStatus[s]).map((s) => `${byStatus[s]} ${s}`).join(' · ') })),
    pinnedCaveat(data.note, 'How these were extracted'),
    data.tables.length
      ? el('ul', { class: 'dt-list' }, ...data.tables.map((t) => el('li', { class: 'dt-item' },
          el('a', { href: `#/doctables?doc=${encodeURIComponent(cur.doc)}&table=${encodeURIComponent(t.document_table_id)}`,
            text: `Page ${t.page_number ?? '—'} · ${num(t.row_count)} × ${num(t.column_count)}` }),
          el('span', { class: `badge ${STATUS_BADGE[t.extraction_status] || 'muted'}`, text: t.extraction_status }),
          t.preview.length ? el('div', { class: 'dt-preview small muted',
            text: t.preview.map((r) => r.join(' | ')).join('  //  ').slice(0, 160) }) : null)))
      : noData('a table in this document')));

  replace(main, page);
  return () => {};
}

File: pipeline/web/doc_tables.py
"""Document table extraction viewer (BETA-099).

Tables detected in parsed documents — the grid the parser produced, its page
context, its extraction status, and a structured (CSV) download built from
exactly what was extracted. Important evidence often sits in tables that
paragraph search and plain-text snippets make hard to read accurately.

The structure comes from `document_tables.table_json` (a JSON array of rows
the parser wrote) — nothing here re-detects a table or reconstructs a cell.
A table whose `table_json` is empty is shown with `extraction_status`
`markdown_only` or `empty`, not with an invented grid. Documents are gated by
the same `DOCUMENT_SEARCH_SOURCES` allowlist as `document_search`.
"""
from __future__ import annotations

import json
import sqlite3

from pipeline.web.public_queries import (
    DOCUMENT_SEARCH_SOURCES,
    _one,
    _public,
    _rows,
)
from pipeline.web.queries import QueryError

_PREVIEW_ROWS = 3
_CONTEXT_ELEMENTS = 2


def _grid(table_json: str | None) -> list[list[str]]:
    try:
        data = json.loads(table_json or "[]")
    except (TypeError, ValueError):
        return []
    if not isinstance(data, list):
        return []
    out: list[list[str]] = []
    for row in data:
        if isinstance(row, list):
            out.append([("" if c is None else str(c)) for c in row])
    return out


def _status(grid: list[list[str]], markdown: str | None) -> str:
    if grid and any(any(cell.strip() for cell in row) for row in grid):
        return "structured"
    if markdown and markdown.strip():
        return "markdown_only"
    return "empty"


_TABLES = ("document_tables", "document_elements", "document_versions",
           "document_records", "evidence_records")


def _doc_guard(conn: sqlite3.Connection, document_id: str) -> dict:
    _public(list(_TABLES))
    row = _one(conn, """
        SELECT d.document_id, d.title, e.source_system, e.source_url,
               e.retrieved_at
        FROM document_records d
        JOIN evidence_records e ON e.evidence_id = d.evidence_id
        WHERE d.document_id = ?""", (document_id,))
    if not row:
        raise QueryError(f"No document {document_id!r}.")
    if row["source_system"] not in DOCUMENT_SEARCH_SOURCES:
        raise QueryError("That document is not available on the portal.")
    return row


def tables(conn: sqlite3.Connection, document_id: str) -> dict:
    doc = _doc_guard(conn, document_id)
    rows = _rows(conn, """
        SELECT dt.document_table_id, dt.document_element_id, dt.row_count,
               dt.column_count, dt.table_json, dt.markdown,
               de.sequence, de.page_number
        FROM document_tables dt
        JOIN document_elements de ON de.document_element_id = dt.document_element_id
        JOIN document_versions v ON v.document_version_id = de.document_version_id
                                 AND v.is_active = 1
        JOIN document_records d ON d.document_id = v.document_id
        WHERE d.document_id = ?
        ORDER BY de.sequence""", (document_id,))

    out = []
    by_status: dict[str, int] = {}
    for r in rows:
        grid = _grid(r["table_json"])
        status = _status(grid, r["markdown"])
        by_status[status] = by_status.get(status, 0) + 1
        out.append({
            "document_table_id": r["document_table_id"],
            "element_id": r["document_element_id"],
            "sequence": r["sequence"],
            "page_number": r["page_number"],
            "row_count": r["row_count"] if r["row_count"] is not None else len(grid),
            "column_count": r["column_count"] if r["column_count"] is not None
            else max((len(row) for row in grid), default=0),
            "extraction_status": status,
            "preview": grid[:_PREVIEW_ROWS],
            "reading_room_link": (f"#/documents?doc={document_id}"
                                   f"&el={r['document_element_id']}"),
        })

    return {
        "document": {"document_id": document_id, "title": doc["title"],
                      "source_url": doc["source_url"],
                      "retrieved_at": doc["retrieved_at"]},
        "tables": out,
        "counts": {"by_status": by_status},
        "statuses": ["structured", "markdown_only", "empty"],
        "note": "Each grid is exactly what the parser wrote to "
                "document_tables — no cell is re-detected or reconstructed. A "
                "markdown_only or empty status means the parse did not produce "
                "a structured grid for that table.",
    }


def table_detail(conn: sqlite3.Connection, document_table_id: str) -> dict:
    _public(list(_TABLES))
    r = _one(conn, """
        SELECT dt.document_table_id, dt.document_element_id, dt.row_count,
               dt.column_count, dt.table_json, dt.markdown,
               de.sequence, de.page_number, de.document_version_id,
               d.document_id, d.title, e.source_system, e.source_url,
               e.retrieved_at
        FROM document_tables dt
        JOIN document_elements de ON de.document_element_id = dt.document_element_id
        JOIN document_versions v ON v.document_version_id = de.document_version_id
        JOIN document_records d ON d.document_id = v.document_id
        JOIN evidence_records e ON e.evidence_id = d.evidence_id
        WHERE dt.document_table_id = ?""", (document_table_id,))
    if not r:
        raise QueryError(f"No table {document_table_id!r}.")
    if r["source_system"] not in DOCUMENT_SEARCH_SOURCES:
        raise QueryError("That document is not available on the portal.")

    grid = _grid(r["table_json"])
    context = _rows(conn, """
        SELECT sequence, element_type, text FROM document_elements
        WHERE document_version_id = ?
          AND sequence BETWEEN ? AND ?
          AND document_element_id <> ?
        ORDER BY sequence""",
        (r["document_version_id"], r["sequence"] - _CONTEXT_ELEMENTS,
         r["sequence"] + _CONTEXT_ELEMENTS, r["document_element_id"]))
    caption = next((c["text"] for c in reversed(context)
                    if c["sequence"] < r["sequence"]
                    and (c["element_type"] or "").upper() in
                    ("HEADING", "TITLE", "CAPTION")), None)

    return {
        "document_table_id": r["document_table_id"],
        "element_id": r["document_element_id"],
        "document": {"document_id": r["document_id"], "title": r["title"],
                      "source_url": r["source_url"],
                      "retrieved_at": r["retrieved_at"]},
        "page_number": r["page_number"],
        "caption": caption,
        "extraction_status": _status(grid, r["markdown"]),
        "grid": grid,
        "markdown": r["markdown"],
        "row_count": len(grid) or r["row_count"],
        "column_count": max((len(row) for row in grid), default=0)
        or r["column_count"],
        "context": [{"element_type": c["element_type"],
                      "text": (c["text"] or "")[:280]} for c in context],
        "reading_room_link": (f"#/documents?doc={r['document_id']}"
                               f"&el={r['document_element_id']}"),
        "note": "The grid is the parser's own extraction. Download it as CSV "
                "to work with exactly those cells; the source document is the "
                "authority for anything the parse got wrong.",
    }
